### PR TITLE
86 get/donors/campagin

### PR DESCRIPTION
**Pull request by DeenaAlghazali**

### CODE REVIEW

File: server/controllers/admin/campaign/getDonorsByCampaignId.ts [hunk 1]
```diff
@@ -0,0 +1,31 @@
...
+
+    const { id } = await paramsSchema.validateAsync(req.params);
+    const donors = await Campaign.findByPk(id, {
+      offset: (page - 1) * 6,
+      limit: 1,
+      include: {
+        model: Donor,
+        attributes: {
+          exclude: ['password', 'createdAt', 'updatedAt'],
+        },
```
Comment (farah2003) [inline server/controllers/admin/campaign/getDonorsByCampaignId.ts:20]: I think we here just the donors for campaigns so no need for campaign attributes so exclude

File: server/controllers/admin/campaign/getDonorsByCampaignId.ts [hunk 2]
```diff
@@ -0,0 +1,31 @@
...
+import { paramsSchema } from '../../../utils/validation';
+
+const getDonorsByCampaignId = async (req: Request, res: Response, next: NextFunction) => {
+  try {
+    const {
+      page = 1,
+    }: any = req.query;
+
+    const { id } = await paramsSchema.validateAsync(req.params);
+    const donors = await Campaign.findByPk(id, {
```
Comment (farah2003) [inline server/controllers/admin/campaign/getDonorsByCampaignId.ts:13]: it is better to use function find and count  it is  return the count of campaigns  that you need in  pagination

**Review by farah2003 — CHANGES_REQUESTED** [review]

File: server/controllers/admin/campaign/getDonorsByCampaignId.ts [hunk 3]
```diff
@@ -0,0 +1,31 @@
+import { Request, Response, NextFunction } from 'express';
+import { Campaign, Donor } from '../../../database/models';
+import { CustomError } from '../../../utils';
+import { paramsSchema } from '../../../utils/validation';
```
Comment (Raghad-Mezied) [inline server/controllers/admin/campaign/getDonorsByCampaignId.ts:4]: They are all from the utils, so  import them from the index

File: server/controllers/admin/campaign/getDonorsByCampaignId.ts [hunk 4]
```diff
@@ -0,0 +1,31 @@
...
+const getDonorsByCampaignId = async (req: Request, res: Response, next: NextFunction) => {
+  try {
+    const {
+      page = 1,
+    }: any = req.query;
+
+    const { id } = await paramsSchema.validateAsync(req.params);
+    const donors = await Campaign.findByPk(id, {
+      offset: (page - 1) * 6,
+      limit: 1,
```
Comment (Raghad-Mezied) [inline server/controllers/admin/campaign/getDonorsByCampaignId.ts:15]: Why the limit is equal to 1?

File: server/routes/adminDonors.ts [hunk 5]
```diff
@@ -0,0 +1,8 @@
+import { Router } from 'express';
+import { getDonorsByCampaignId } from '../controllers';
+
+const adminDonors = Router();
+
+adminDonors.get('/campaign/donors/:id', getDonorsByCampaignId);
```
Comment (Raghad-Mezied) [inline server/routes/adminDonors.ts:6]: Is it allowed for anyone to request this endpoint? 
And you need to fix the test depend on that

File: server/controllers/admin/campaign/getDonorsByCampaignId.ts [hunk 6]
```diff
@@ -0,0 +1,31 @@
...
+      include: {
+        model: Donor,
+        attributes: {
+          exclude: ['password', 'createdAt', 'updatedAt'],
+        },
+      },
+    });
+    if (!donors) throw new CustomError('Campaign does not exist', 400);
```
Comment (Raghad-Mezied) [inline server/controllers/admin/campaign/getDonorsByCampaignId.ts:23]: If there are no doners then nothing wrong with the request but it is just an empty data

**Review by Raghad-Mezied — CHANGES_REQUESTED** [review]

File: server/tests/route.test.ts [hunk 7]
```diff
@@ -426,6 +426,45 @@ describe('GET/campaigns', () => {
     expect(response.body.data.campaigns).toEqual([]);
   });
 });
+describe('get admin/campaign/donor/:campaignId', () => {
+  test('get data successfully', async () => {
+    const campaignId = 3;
+    const donor = {
+      id: 4,
+      name: 'sami',
+      email: 'sami@gmail.com',
```
Comment (Raghad-Mezied) [inline server/tests/route.test.ts:435]: Test the case of an unauthorized user.

**Review by Raghad-Mezied — CHANGES_REQUESTED** [review]

**Review by Raghad-Mezied — APPROVED** [review]

**Review by Raghad-Mezied — APPROVED** [review]

File: server/routes/adminDonors.ts [hunk 8]
```diff
@@ -0,0 +1,9 @@
+import { Router } from 'express';
+import { getDonorsByCampaignId } from '../controllers';
+import { authUser, authAdmin } from '../middlewares';
+
+const adminDonors = Router();
+
+adminDonors.get('/campaign/donors/:id', authUser, authAdmin, getDonorsByCampaignId);
```
Comment (Mohammed-Hamada) [inline server/routes/adminDonors.ts:7]: Here, i think if endpoint be like `/campaign/:id/donors` that would be more clear to read and understand

File: server/controllers/admin/campaign/getDonorsByCampaignId.ts [hunk 9]
```diff
@@ -0,0 +1,27 @@
...
+      limit: 6,
+      include: {
+        model: Campaign,
+        where: { id },
+      },
+      attributes: {
+        exclude: ['password', 'createdAt', 'updatedAt'],
+      },
+    });
+    res.json({ message: 'Success', data: donors });
```
Comment (Mohammed-Hamada) [inline server/controllers/admin/campaign/getDonorsByCampaignId.ts:20]: `data: { donors }` better than `data: donors`

**Review by Mohammed-Hamada — CHANGES_REQUESTED** [review]

File: server/controllers/admin/campaign/getDonorsByCampaignId.ts (hunk 2 above)
Comment (Mohammed-Hamada) [reply server/controllers/admin/campaign/getDonorsByCampaignId.ts]: https://sequelize.org/docs/v6/core-concepts/model-querying-finders/#findandcountall

File: server/controllers/admin/campaign/getDonorsByCampaignId.ts [hunk 10]
```diff
@@ -0,0 +1,27 @@
...
+  } catch (error) {
+    if (error.name === 'ValidationError') next(new CustomError(error.details[0].message, 400));
```
Comment (Mohammed-Hamada) [inline server/controllers/admin/campaign/getDonorsByCampaignId.ts:22]: just you want to send `error.message`, that equals `error.details[0].message`

**Review by Mohammed-Hamada — CHANGES_REQUESTED** [review]

**Review by farah2003 — APPROVED** [review]

**Review by Mohammed-Hamada — APPROVED** [review]

**Review by HaniOlwan — APPROVED** [review]

**Pull request merged**